build.go

```diff
@@ -12,93 +11,54 @@ import (
 
 //go:generate faux --interface InstallProcess --output fakes/install_process.go
 type InstallProcess interface {
-	Execute(workingDir, layerPath string) error
-}
-
-//go:generate faux --interface Calculator --output fakes/calculator.go
-type Calculator interface {
-	Sum(paths ...string) (string, error)
+	ShouldRun(layer packit.Layer, workingDir string) (should bool, checksum string, rubyVersion string, err error)
+	Execute(workingDir, layerPath string, config map[string]string) error
 }
 
 //go:generate faux --interface EntryResolver --output fakes/entry_resolver.go
 type EntryResolver interface {
-	Resolve(string, []packit.BuildpackPlanEntry, []interface{}) (packit.BuildpackPlanEntry, []packit.BuildpackPlanEntry)
 	MergeLayerTypes(string, []packit.BuildpackPlanEntry) (launch, build bool)
 }
 
-//go:generate faux --interface VersionResolver --output fakes/version_resolver.go
-type VersionResolver interface {
-	Lookup() (version string, err error)
-	CompareMajorMinor(string, string) (bool, error)
-}
-
-func Build(
-	installProcess InstallProcess,
-	calculator Calculator,
-	logger LogEmitter,
-	clock chronos.Clock,
-	entries EntryResolver,
-	versionResolver VersionResolver,
-) packit.BuildFunc {
+func Build(installProcess InstallProcess, logger LogEmitter, clock chronos.Clock, entries EntryResolver) packit.BuildFunc {
 	return func(context packit.BuildContext) (packit.BuildResult, error) {
 		logger.Title("%s %s", context.BuildpackInfo.Name, context.BuildpackInfo.Version)
 
-		entry, _ := entries.Resolve("gems", context.Plan.Entries, []interface{}{})
-
 		gemsLayer, err := context.Layers.Get(LayerNameGems)
 		if err != nil {
 			return packit.BuildResult{}, err
 		}
 
-		rubyVersion, err := versionResolver.Lookup()
+		should, sum, rubyVersion, err := installProcess.ShouldRun(gemsLayer, context.WorkingDir)
 		if err != nil {
 			return packit.BuildResult{}, err
 		}
 
-		cachedRubyVersion, cachedVersionExists := gemsLayer.Metadata["ruby_version"].(string)
-		// If there is no cached ruby version, then the match is true
-		rubyVersionMatch := true
-
-		if cachedVersionExists {
-			// rubyVersionMatch will be true if at least the major and minor versions match
-			rubyVersionMatch, err = versionResolver.CompareMajorMinor(cachedRubyVersion, rubyVersion)
-			if err != nil {
-				return packit.BuildResult{}, err
-			}
-		}
+		if !should {
+			logger.Process("Reusing cached layer %s", gemsLayer.Path)
+			logger.Break()
 
-		var sum string
-		_, err = os.Stat(filepath.Join(context.WorkingDir, "Gemfile.lock"))
-		if err != nil {
-			if !os.IsNotExist(err) {
-				return packit.BuildResult{}, fmt.Errorf("failed to stat Gemfile.lock: %w", err)
-			}
-		} else {
-			sum, err = calculator.Sum(filepath.Join(context.WorkingDir, "Gemfile"), filepath.Join(context.WorkingDir, "Gemfile.lock"))
+			err := os.RemoveAll(filepath.Join(context.WorkingDir, ".bundle", "config"))
```

Comment:
Why do we need to remove `.bundle/config` when rebuilding from a cached layer?

Reply:
The `.bundle/config` that is removed here is the one in the `/workspace`. This should no longer be in the `/workspace` in either the case that we run the build process or that we reuse a layer. Since it comes from the user's source code, it could be present in the case that we choose to reuse the layer, and so we need to remove it so that we end up with the same kind of resulting filesystem as we would if we had run the build process.